blocks/serialization.py

```diff
@@ -1,65 +1,181 @@
+        ``True``.
+    allow_duplicates : bool, optional
+        Allow multiple shared variables to have the same name, in which
+        case they will be numbered e.g. `x`, `x_2`, `x_3`, etc. Defaults to
+        ``True``.
+
+    Raises
+    ------
+    ValueError
+        If an unnamed shared variable is encountered and `allow_unnamed` is
+        ``False``, or if two shared variables have the same name, and
+        `allow_duplicates` is ``False``.
 
-Python is unable to serialize nested functions. Make sure that all your \
-functions are available in the global namespace."""
+    """
+    def __init__(self, zip_file, allow_unnamed=True, allow_duplicates=True):
+        super(PersistentParameterID, self).__init__(zip_file)
+        self.name_counter = defaultdict(int)
+        self.ndarray_names = {}
+        self.allow_unnamed = allow_unnamed
+        self.allow_duplicates = allow_duplicates
+
+    def _resolve_name(self, obj):
+        if id(obj) in self.ndarray_names:
+            name = self.ndarray_names[id(obj)]
+            count = self.name_counter[name]
+            if count:
+                if not self.allow_duplicates:
+                    raise ValueError("multiple shared variables with the name "
+                                     "`{0}` found".format(name))
+                name = '{0}_{1}'.format(name, count + 1)
+            self.name_counter[name] += 1
+            return name
+        return super(PersistentParameterID, self)._resolve_name(obj)
+
+    def __call__(self, obj):
+        if isinstance(obj, SharedVariable):
+            if obj.name:
+                if obj.name == 'pkl':
+                    raise ValueError(
+                        "can't pickle shared variable with name `pkl`")
+                if hasattr(obj.tag, 'annotations'):
+                    name = BRICK_DELIMITER.join(
+                        [brick.name for brick
+                         in obj.tag.annotations[0].get_unique_path()] +
+                        [obj.name])
+                else:
+                    name = obj.name
+                self.ndarray_names[id(obj.container.storage[0])] = name
+            elif not self.allow_unnamed:
+                raise ValueError("unnamed shared variable, {}".format(obj))
+        return super(PersistentParameterID, self).__call__(obj)
+
+
+class PicklerWithWarning(pickle.Pickler):
+    if six.PY2:
+        dispatch = pickle.Pickler.dispatch.copy()
+    else:
+        dispatch_table = copyreg.dispatch_table.copy()
+
+    def save_global(self, obj, **kwargs):
+        module = getattr(obj, "__module__", None)
+        if module == '__main__':
+            warnings.warn(
+                MAIN_MODULE_WARNING.format(kwargs.get('name', obj.__name__))
+            )
+        pickle.Pickler.save_global(self, obj, **kwargs)
+    if six.PY3:
+        dispatch_table[six.types.FunctionType] = save_global
+    else:
+        dispatch[six.types.ClassType] = save_global
```

Comment:
Oh, and since cPickle doesn't exist in Python 3 anymore, do you have
timings for that as well?

On Tue, May 19, 2015 at 8:46 AM, Bart van Merriënboer <
bart.vanmerrienboer@gmail.com> wrote:

> Did you try timing it without all the usual empty dictionaries in the log?
> Curious to see the impact of that.
> On May 19, 2015 08:44, "Dmitry Bogdanov" <notifications@github.com> wrote:
>
>> In blocks/serialization.py
>> <https://github.com/mila-udem/blocks/pull/615#discussion_r30594588>:
>>
>> > +    if six.PY2:
>> > +        dispatch = pickle.Pickler.dispatch.copy()
>> > +    else:
>> > +        dispatch_table = copyreg.dispatch_table.copy()
>> > +
>> > +    def save_global(self, obj, **kwargs):
>> > +        module = getattr(obj, "__module__", None)
>> > +        if module == '__main__':
>> > +            warnings.warn(
>> > +                MAIN_MODULE_WARNING.format(kwargs.get('name', obj.__name__))
>> > +            )
>> > +        pickle.Pickler.save_global(self, obj, **kwargs)
>> > +    if six.PY3:
>> > +        dispatch_table[six.types.FunctionType] = save_global
>> > +    else:
>> > +        dispatch[six.types.ClassType] = save_global
>>
>> I apologize for writing so many posts instead of one. I checked out your
>> branch and benchmarked your dump and load:
>>
>> In [11]: %timeit  dump(log, open('log.zip', 'w'))1 loops, best of 3: 1min 38s per loop
>>
>> In [12]: %timeit  cPickle.dump(log, open('log.pkl', 'w'))1 loops, best of 3: 3.3 s per loop
>>
>> In [14]: %timeit  load(open('log.zip'))1 loops, best of 3: 2min 10s per loop
>>
>> In [15]: %timeit  cPickle.load(open('log.pkl'))1 loops, best of 3: 2.83 s per loop
>>
>> Also, I read more carefully your code, and it seems like you use your
>> custom dump also for save_separately elements. Which effectively means
>> that if this PR is merged, loading my logs will take up to 2 minutes
>> instead of a few seconds.
>>
>> That being said, I still like the PR. But let's figure out how to reach
>> its goals without drastic slowdowns.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/mila-udem/blocks/pull/615/files#r30594588>.
>>
>